ext/drmgr/drmgr.c

```diff
@@ -594,6 +645,104 @@ cblist_delete_local(void *drcontext, cb_list_t *l, size_t local_num)
     } /* else nothing to do */
 }
 
+/* Creates a global copy of a cb list. Use cblist_delete() to destroy. */
+static void
+cblist_create_global(cb_list_t *src, cb_list_t *dst)
+{
+    dst->capacity = src->capacity;
+    dst->cbs.array = dr_global_alloc(src->capacity * src->entry_sz);
+    cblist_copy(src, dst);
+}
+
+/***************************************************************************
+ * OPCODE INSTRUM HASHTABLE
+ */
+
+static void
+drmgr_destroy_opcode_cb_list(void *opcode_cb_list_opaque)
+{
+    cb_list_t *opcode_cb_list = (cb_list_t *)opcode_cb_list_opaque;
+    cblist_delete(opcode_cb_list);
+    dr_global_free(opcode_cb_list, sizeof(cb_list_t));
+}
+
+static void
+drmgr_init_opcode_hashtable(hashtable_t *opcode_instrum_table)
+{
+    hashtable_init_ex(opcode_instrum_table, 8, HASH_INTPTR, false, false,
+                      drmgr_destroy_opcode_cb_list, NULL, NULL);
+}
+
+/* Returns false if opcode instrumentation is not applicable, i.e., no registration.
+ */
+static bool
+drmgr_set_up_local_opcode_table(IN instrlist_t *bb, IN cb_list_t *insert_list,
+                                INOUT hashtable_t *local_opcode_instrum_table)
+{
+    instr_t *inst, *next_inst;
+    int opcode;
+    bool is_opcode_instrum_applicable = false;
+
+    dr_rwlock_read_lock(opcode_table_lock);
+
+    /* A high-level check to avoid expensive allocation and locking bb_cb_lock.
+     * In particular, we iterate over the bb and check whether any applicable opcode
+     * event is registered.
+     */
+    for (inst = instrlist_first(bb); inst != NULL; inst = next_inst) {
+        next_inst = instr_get_next(inst);
+        if (!instr_opcode_valid(inst))
+            continue;
+        opcode = instr_get_opcode(inst);
+        cb_list_t *opcode_cb_list =
+            hashtable_lookup(&global_opcode_instrum_table, (void *)(intptr_t)opcode);
+        if (opcode_cb_list != NULL) {
+            is_opcode_instrum_applicable = true;
+            break;
+        }
+    }
+
+    /* Just return if no opcode event is registered for any instruction in the BB. */
+    if (!is_opcode_instrum_applicable) {
+        dr_rwlock_read_unlock(opcode_table_lock);
+        return false;
+    }
+
+    /* Since both opcode and insert events are handled during stage 3, they need to be
+     * jointly organized according to their priorities. This is done now.
+     *
+     * An alternative approach is to do the merging upon registration of opcode
+     * events, and updating all mapped cb lists of opcodes when further bb insert events
+     * are registered/unregistered. However, in order to avoid races, this would require
+     * the dr bb event processed by drmgr to create a local copy of the entire opcode
+     * table every time.
```

Comment:
Hmm, I guess I'm imagining there only being a couple of opcode instru cbs, and we'd just put them on the same cblist_instrumentation as the regular instruction insertion cbs, and in the existing loop over the instrs (what you have now as drmgr_bb_event_do_insertion_per_instr) we'd just check the opcode vs the opcode stored with the cb before deciding whether to invoke the cb.  No need for a hashtable.  It seems a simpler and more efficient solution?  Though I only thought it about for a few minutes.  Unless you're anticipating a whole lot of these opcode cbs?

Reply:
hmm. So in the PR comment I said:

> I avoided using one monolithic cb list as clients with a long list of different registered opcode events would have required drmgr to do wasteful iterations.

It depends on how many opcode events are registered. I guess I can revert and make the code simpler.

If the user has many opcode events, then I guess the user should refrain from using this interface and use a large switch statement (something similar to hiw dr memory maintains def tags). 


If there are many, they I would recommend the

Reply:
I'm ok with it as is.

Reply:
Okay. I'm fine with both approaches tbh because they both offer their own benefits. If this is the one we are going for, then I'll add a comment on the other.